samples/qtask_thread.c

```diff
@@ -0,0 +1,344 @@
+    pthread_cond_t stopcond = PTHREAD_COND_INITIALIZER;
+    pthread_t thread = 0;
+    td_orderedwrite twritedata = {0};
+    hts_tpool *pool = NULL;
+    hts_tpool_process *queue1 = NULL, *queue2 = NULL;
+
+    //split_t3 infile threadcount outdir
+    if (argc != 4) {
+        print_usage(stdout);
+        goto end;
+    }
+    inname = argv[1];
+    cnt = atoi(argv[2]);
+    outdir = argv[3];
+
+    if (cnt < 1) {
+        cnt = 1;
+    }
+
+    //allocate space for output
+    size = sizeof(char) * (strlen(outdir) + sizeof("/1.sam") + 1); //space for output file name and null termination
+    file1 = malloc(size); file2 = malloc(size);
+    if (!file1 || !file2) {
+        printf("Failed to set output path\n");
+        goto end;
+    }
+
+    //output file names
+    snprintf(file1, size, "%s/1.sam", outdir);  //for SAM output
+    snprintf(file2, size, "%s/2.bam", outdir);  //for BAM output
+    //bam data storage
+    if (!(bamdata = bam_init1())) {
+        printf("Failed to initialize bamdata\n");
+        goto end;
+    }
+    //thread pool
+    if (!(pool = hts_tpool_init(cnt))) {
+        printf("Failed to create thread pool\n");
+        goto end;
+    }
+    //queue to use with thread pool, no queuing of results for Q1 and results queued for Q2
+    if (!(queue1 = hts_tpool_process_init(pool, cnt * 2, 1)) || !(queue2 = hts_tpool_process_init(pool, cnt * 2, 0))) {
+        printf("Failed to create queue\n");
+        goto end;
+    }
+    //open input file - r reading
+    if (!(infile = sam_open(inname, "r"))) {
+        printf("Could not open %s\n", inname);
+        goto end;
+    }
+
+    //open output files - w write as SAM, wb  write as BAM
+    outfile1 = sam_open(file1, "w"); outfile2 = sam_open(file2, "wb");
+    if (!outfile1 || !outfile2) {
+        printf("Could not open output file\n");
+        goto end;
+    }
+    //read header, required to resolve the target names to proper ids
+    if (!(in_samhdr = sam_hdr_read(infile))) {
+        printf("Failed to read header from file!\n");
+        goto end;
+    }
+    //start output writer thread for ordered processing
+    twritedata.outfile = outfile2; twritedata.queue = queue2; twritedata.done = &stopcond; twritedata.lock = &stopcondsynch;
+    twritedata.samhdr = in_samhdr;
+    if (pthread_create(&thread, NULL, threadfn_orderedwrite, &twritedata)) {
+        printf("Failed to create writer thread\n");
+        goto end;
+    }
+    //write header
+    if ((sam_hdr_write(outfile1, in_samhdr) == -1) || (sam_hdr_write(outfile2, in_samhdr) == -1)) {
+        printf("Failed to write header\n");
+        goto end;
+    }
+    //check flags and schedule
+    while ((c = sam_read1(infile, in_samhdr, bamdata)) >= 0) {
+        if (bamdata->core.flag & BAM_FREAD1) {
+            //read1, scheduling for unordered output, using Q1; processing occurs in order of thread execution
+            td_unordered *tdata = calloc(1, sizeof(td_unordered));
```

Comment:
I appreciate it's a trivial example, but it's not a sensible one to do in practice.

This may be fine for demonstration purposes, but only if it has comments in both the code and DEMO.md pointing out this is a noddy example for educational purposes only and not a recommended way of doing things.  I fear otherwise people may look at this example and think that dispatching thread work loads on a read-by-read basis is an appropriate thing to do.  Our threading mechanism is just too heavyweight for that (we'd need to buffer up multiple tasks and do them in batches).

If we wish to go for a very simple and easy to understand example, it'd be better to completely avoid the rest of htslib APIs and anything with biological data, rather than suggesting something inappropriate.  Take a look at the end of thread_pool.c for some worked examples instead.

If we want it relevant to sequencing data, then I still think we ideally need a different scenario.

I still think splitting by region is a better example as it can come with a narrative educating people about the features.  Eg.

1. Read a bed file (or just a file of region names read using `fgets`).
2. Per region, dispatch a task per region.  Each task creates a new region specific output file, so we can discuss here the difference between ordered (where we have a tangible result and it must occur in order) and unordered (task has done something during its execution, and the "result" is just a status code).
3. Further discussion on the inefficiencies, left as an exercise to the reader.  Eg consider how this works when we have many tiny regions.  We may wish to amalgamate multiple regions into a single task.

Reply:
[bambi](https://github.com/wtsi-npg/bambi) is the only thing I can think of that uses a no-results queue to [decompress batches of bcl files](https://github.com/wtsi-npg/bambi/blob/4c25263f0/src/i2b.c#L1581).

A realistic example for use here might be something that calculates a statistic over all the reads that doesn't depend on the order (e.g. read base type counts, or a quality histogram).  For efficiency it should batch up some reads before feeding them into the thread pool job.  Each batch job would need to merge its counts into a global structure, which would need to be locked.  Or possibly it could implement a multi-level merge strategy with the final stage at the end, so that the thread jobs are less likely to get stuck waiting on mutexes.

Reply:
removed this sample and 2 different ones added. 
ordered processing updated an aux tag and saves the same.
unordered one splits the file file specific for given regions.